runtime/runtime/src/bandwidth_scheduler/scheduler.rs

```diff
@@ -0,0 +1,765 @@
+//!
+//! TODO(bandwidth_scheduler) - think/prove/test if this is enough to handle missing chunks. We
+//! actually know if other chunks in the same block are missing, which is pretty powerful. I need to
+//! think more about it. Maybe there was no fatal flaw in the previous design after all? ;-;
+//!
+//! Future improvements: It's a bit wasteful to not send anything when a chunk is missing. We could
+//! track how much a shard already sent to the shard with missing chunks at previous heights, and
+//! allow to send receipts to it, but make sure that their total size doesn't exceed the grant that
+//! was given when the chunk wasn't missing. This would improve bandwidth utilization in scenarios
+//! with many missing chunks.
+//!
+//! ## Resharding
+//!
+//! During resharding the set of active shards changes. For example, before resharding there could
+//! be three shard ids: [2, 4, 6], but during resharding shard 6 could be split into 7 and 8, and
+//! the resulting list of shard ids would be: [2, 4, 7, 8]. From bandwidth scheduler's perspective,
+//! the only moment that's really problematic is the resharding boundary. For one block height the
+//! sender shards will be from the old layout, but receiver shards will be from the new one. Ideally
+//! bandwidth scheduler would handle this gracefully, but for now this is not implemented. The
+//! bandwidth grants will be slightly wrong for one height, but it is not the end of the world. At
+//! the next height the shard will fully switch to the new layout and all the grants will be correct
+//! again.
+//!
+//! To properly handle resharding we would have to:
+//! * Properly set the `senders` and `receivers` on a resharding boundary.
+//! * Interpret bandwidth requests using the BandwidthSchedulerParams that they were created with
+//! * Make sure that `BandwidthSchedulerParams` are correct on the resharding boundary
+//!
+//! It's doable, but it's out of scope for the initial version of the scheduler. There was a bit of
+//! an effort to write the code in a future-proof way, so adding full support for resharding
+//! shouldn't be too painful.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::rc::Rc;
+
+use near_primitives::bandwidth_scheduler::{
+    Bandwidth, BandwidthRequest, BandwidthRequestValues, BandwidthRequests,
+    BandwidthSchedulerParams, BandwidthSchedulerState, BlockBandwidthRequests, LinkAllowance,
+};
+use near_primitives::types::ShardId;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use super::shard_mapping::{
+    SchedulerShardIndex, SchedulerShardIndexMap, SchedulerShardLinkMap, SchedulerShardMapping,
+    ShardLink,
+};
+
+/// How many bytes of outgoing receipts can be sent from one shard to another at the current height.
+/// Produced by the bandwidth scheduler.
+pub struct GrantedBandwidth {
+    granted: BTreeMap<(ShardId, ShardId), Bandwidth>,
```

Comment:
I don't really understand the desire to remove `Ord` from `ShardId`. It's an ID, it's pretty intuitive that it can be ordered. And it's very useful for things like deduplication.

I can change it to a HashMap, but I don't like HashMaps becuase they have a potential for causing nondeterminism :/